go/store/prolly/tree_merge.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prolly
+
+import (
+	"bytes"
+	"context"
+	"io"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+const patchBufferSize = 1024
+
+// TupleMergeFn is a callback that handles 3-way merging of tuples.
+// A typical implementation will attempt a cell-wise merge of the tuples,
+// or register a conflict if a such a merge is not possible.
+type TupleMergeFn func(left, right Diff) (Diff, bool)
+
+// ThreeWayMerge implements a three-way merge algorithm using |base| as the common ancestor, |right| as
+// the source branch, and |left| as the destination branch. Both |left| and |right| are diff'd against
+// |base| to compute merge patches, but rather than applying both sets of patches to |base|, patches from
+// |right| are applied directly to |left|. This reduces the amount of write work and improves performance.
+// In the case that a key-value pair was modified on both |left| and |right| with different resulting
+// values, the TupleMergeFn is called to perform a cell-wise merge, or to throw a conflict.
+func ThreeWayMerge(ctx context.Context, left, right, base Map, cb TupleMergeFn) (final Map, err error) {
+	ld, err := treeDifferFromMaps(ctx, base, left)
+	if err != nil {
+		return Map{}, err
+	}
+
+	rd, err := treeDifferFromMaps(ctx, base, right)
+	if err != nil {
+		return Map{}, err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	buf := newPatchBuffer(patchBufferSize)
+
+	// iterate |ld| and |rd| in parallel, populating |buf|
+	eg.Go(func() (err error) {
+		defer func() {
+			if cerr := buf.close(); err != nil {
```

Comment:
I think this should be if `err == nil` so it only overwrites the error when there is no error already existing.  This currently overwrites the error only if there is a pre-existing error.

Reply:
dangerous!